src/cpp/rtps/builtin/discovery/database/DiscoveryDataBase.cpp

```diff
@@ -34,9 +34,88 @@ DiscoveryDataBase::DiscoveryDataBase(
         fastrtps::rtps::GuidPrefix_t server_guid_prefix)
     : server_guid_prefix_(server_guid_prefix)
     , server_acked_by_all_(false)
+    , enabled_(true)
 {
 }
 
+DiscoveryDataBase::~DiscoveryDataBase()
+{
+    if (!clear().empty())
+    {
+        logError(DISCOVERY_DATABASE, "Destroying a NOT cleared database");
+    }
+}
+
+std::vector<fastrtps::rtps::CacheChange_t*> DiscoveryDataBase::clear()
+{
+    std::unique_lock<share_mutex_t> lock(sh_mtx_);
+
+    // Cannot clear an enabled database, since there could be inconsistencies after the process
+    if (enabled_)
+    {
+        logError(DISCOVERY_DATABASE, "Cannot clear an enabled database");
+        return std::vector<fastrtps::rtps::CacheChange_t*>({});
+    }
+    logInfo(DISCOVERY_DATABASE, "Clearing DiscoveryDataBase");
+
+    /* Clear receive queues */
+    pdp_data_queue_.Clear();
+    edp_data_queue_.Clear();
+
+    /* Clear by_topic collections */
+    writers_by_topic_.clear();
+    readers_by_topic_.clear();
+
+    /* Clear list of dirty topics */
+    dirty_topics_.clear();
+
+    /* Clear disposals list */
+    disposals_.clear();
+
+    /* Clear to_send collections */
+    pdp_to_send_.clear();
+    edp_publications_to_send_.clear();
+    edp_subscriptions_to_send_.clear();
+
+    /* Unmatch and clear writers_ */
+    for (auto writers_it = writers_.begin(); writers_it != writers_.end();)
+    {
+        unmatch_writer_(writers_it->first);
+        writers_it = delete_writer_entity_(writers_it);
+    }
+
+    /* Unmatch and clear readers_ */
+    for (auto readers_it = readers_.begin(); readers_it != readers_.end();)
+    {
+        unmatch_reader_(readers_it->first);
+        readers_it = delete_reader_entity_(readers_it);
+    }
+
+    /* Unmatch and clear participants_ */
+    // Unmatching a participant also changes the ACK status on the relevant participants, so all participants need to be
+    // unmatched first to avoid unmatching already deleted participants
+    for (auto participant: participants_)
+    {
+        unmatch_participant_(participant.first);
+    }
+    // Once all partiicipants have been unmatched, the it is safe to delete them
+    for (auto participants_it = participants_.begin(); participants_it != participants_.end();)
+    {
+        participants_it = delete_participant_entity_(participants_it);
+    }
+
+    /* Reset state parameters */
+    server_acked_by_all_ = false;
+    local_servers_count_ = 0;
+
+    /* Clear changes to release */
+    std::vector<fastrtps::rtps::CacheChange_t*> changes_to_release = changes_to_release_;
+    changes_to_release_.clear();
+
+    /* Return the collection of changes that are no longer owned by the database */
+    return changes_to_release;
+}
+
 bool DiscoveryDataBase::pdp_is_relevant(
         const eprosima::fastrtps::rtps::CacheChange_t& change,
         const eprosima::fastrtps::rtps::GUID_t& reader_guid) const
```

Comment:
It does not look like it, but we must be sure that is_relevant (and the rest of public API) will not break the program in a disabled state.